### PR TITLE
Apply Finality to blocks prior to L1 Finality Singal

### PR DESCRIPTION
**Pull request by axelKingsley**
When L1 processing on a node lags far behind the L1 processing of the Supervisor, the L1 Finality signal (which we only store the latest of) can't be used to determine L2 finality.

To fix that -- if the L1 finality signal is more recent than the L1 block the node is currently using, we can just deduce that the L2 blocks being derived currently are *already* finalized. This PR updates the `Finalized` function of ChainsDB to make such a determination.

# Testing
This issue was diagnosed via a local Kurtosis Devnet, where L2 finality was updating for a while, but stopped and never came back.

This is manually tested by observing a Kurtosis Devnet on this branch, which updates the Finalized L2 consistently.

### CODE REVIEW

File: op-supervisor/supervisor/backend/db/query.go [hunk 1]
```diff
@@ -171,6 +171,23 @@ func (db *ChainsDB) Finalized(chainID types.ChainID) (types.BlockSeal, error) {
...
+		return types.BlockSeal{}, types.ErrUnknownChain
+	}
+	latestDerivedFrom, latestDerived, err := xDB.Latest()
+	if err != nil {
+		return types.BlockSeal{}, fmt.Errorf("could not get the latest derived pair for chain %s: %w", chainID, err)
+	}
+	// if the finalized L1 block is newer than the latest L1 block used to derive L2 blocks,
+	// the finality signal automatically applies to all previous blocks, including the latest derived block
+	if finalizedL1.Number > latestDerivedFrom.Number {
+		return latestDerived, nil
```
Comment (protolambda) [inline op-supervisor/supervisor/backend/db/query.go]: Technically this is unsafe, when the cross-safe DB is on a different L1 chain than the L1 finality signal.
Adding a log about assuming finality in this code-path would be nice.

**Review by protolambda — COMMENTED** [review]
Change looks good, but needs a test. An action test with a verifier node that receives the finality signal before the L2 block can capture this edge-case.

File: op-supervisor/supervisor/backend/db/query.go (hunk 1 above)
Comment (axelKingsley) [reply op-supervisor/supervisor/backend/db/query.go]: I'm adding a `Warn`, but I wonder if this is actually safe --

> when the cross-safe DB is on a different L1 chain than the L1 finality signal

The `latestDerived` is *definitely* finalized, because the supervisor database is consistent to itself. It is *kinda* possible for the `op-node` to be on a different L1 chain, but if it deviated, all sorts of prior signals should reveal that the node is inconsistent. It should always recognize this finalized L2 block because if it didn't it would be running diverged L1.

For `cross-safe DB` to be on a different L1 than the `L1 finality signal`, they'd have to be sourced from different L1 connections, but they are not.

Maybe this is unsafe for Standard nodes who are just doing the read API? 🤔 🤔 🤔

**Comment by axelKingsley** [comment]
Discussed offline -- we would like to merge this PR for the following reasons:

- Fixing the finality signal is a critical unblock for Kurtosis
- I tested this manually and observe the correct improvement of finality, where previously Finality fell behind
- The Action tests (where we wanted to add testing) is very heavy for Supervisor/Node interactions right now and can't be simply built.

We are going to prioritize improving the event system to support easier Action tests, and then will prioritize new Action testing, including tests for finality delivery like this.

**Review by protolambda — APPROVED** [review]
